### ISSUE TITLE
O_NONBLOCK enabled by default for IO.pipe used to capture subprocess stdout on Java 21

### ISSUE THREAD

**Issue by kschoelhorn** [jruby/jruby#8069]
## Environment Information
  
* JRuby 9.4.5.0 on Java 21.0.1, no JRUBY_OPTS
  `jruby 9.4.5.0 (3.1.4) 2023-11-02 1abae2700f OpenJDK 64-Bit Server VM 21.0.1+12-LTS on 21.0.1+12-LTS +jit [x86_64-linux]`
* RHEL 9.2 with Linux 5.14.0-284.25.1.el9_2.x86_64

## Expected Behavior

Capturing the stdout or stderr of a subprocess with `Open3.capture` or by manually creating an `IO.pipe` and passing it to `spawn` should create a blocking stdout fd for the subprocess, i.e. O_NONBLOCK should _not_ be enabled.

## Actual Behavior

Running JRuby with Java 21.0.1 creates a stdout with O_NONBLOCK  enabled, which leads to problems if the subprocess wants to execute a larger `write` and is not expecting O_NONBLOCK (which is usually not the default for stdout). E.g. it causes a `BlockingIOError: [Errno 11] write could not complete without blocking` error in python3.

Using backticks or the `java.lang.ProcessBuilder` do not show the issue, only `IO.pipe`+`spawn` and `Open3` (which also seems to use `IO.pipe`+`spawn` internally).

I tested with both JRuby 9.4.5.0 and 9.2.21.0 and both show the same behavior. Java 17 and below do not show this issue and CRuby is also working fine:

```
$ ruby capturePython.rb
CRuby 3.0.3
Open3.capture2: stdout io flags: 1 == O_WRONLY
Manual spawn: stdout io flags: 1 == O_WRONLY
Backticks: stdout io flags: 1 == O_WRONLY

$ JAVA_HOME=/usr/lib/jvm/java-17 jruby capturePython.rb
JRuby 9.4.5.0, Java 17.0.6
Open3.capture2: stdout io flags: 1 == O_WRONLY
Manual spawn: stdout io flags: 1 == O_WRONLY
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY

$ JAVA_HOME=/usr/lib/jvm/java-21 jruby capturePython.rb
JRuby 9.4.5.0, Java 21.0.1
Open3.capture2: stdout io flags: 2049 == O_WRONLY+O_NONBLOCK
Manual spawn: stdout io flags: 2049 == O_WRONLY+O_NONBLOCK
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY
```

## Reproducer

`capturePython.rb`
```ruby
require 'open3'

if defined? JRUBY_VERSION
  puts "JRuby #{JRUBY_VERSION}, Java #{java.lang.System.getProperty('java.version')}"
else
  puts "CRuby #{RUBY_VERSION}"
end

output, _status = Open3.capture2 './printFlags.py'
puts "Open3.capture2: #{output}"

read, write = IO.pipe
pid = spawn("./printFlags.py", :out => write)
write.close
puts "Manual spawn: #{read.read}"
Process.wait pid

puts "Backticks: #{`./printFlags.py`}"

if defined? JRUBY_VERSION
  p = java.lang.ProcessBuilder.new("./printFlags.py").start()
  p.getOutputStream().close()
  puts("java.lang.ProcessBuilder: #{java.lang.String.new(p.getInputStream().readAllBytes())}")
  p.waitFor()
end
```

`printFlags.py`
```python
#!/usr/bin/env python3

import fcntl
import os
import sys

def decode_flags(flags):
    flags_to_check = {
        "O_RDONLY": os.O_RDONLY,
        "O_WRONLY": os.O_WRONLY,
        "O_RDWR": os.O_RDWR,
        "O_APPEND": os.O_APPEND,
        "O_NONBLOCK": os.O_NONBLOCK,
    }

    result = []
    for name, flag in flags_to_check.items():
        if flags & flag:
            result.append(name)
            flags &= ~flag
    
    if flags:
        result.append(str(flags))
    
    return "+".join(result)

stdout_io_flags = fcntl.fcntl(sys.stdout.fileno(), fcntl.F_GETFL)
print(f"stdout io flags: {stdout_io_flags} == {decode_flags(stdout_io_flags)}")
```

**Comment by headius** [comment]
Wow that is a weird thing to break in Java 21.

**Comment by headius** [comment]
I can see nothing we are doing that would be different on 21, so I'm investigating whether something changed in the JDK to make pipe streams nonblocking by default.

**Comment by headius** [comment]
Step debugging shows me that the channel and the JRuby wrappers around it do indeed have "nonblock = false" up to the point the pipe streams are returned. Possibly the spawn call is causing them to become nonblock?

**Comment by kschoelhorn** [comment]
I tested it with the jdk binaries from https://jdk.java.net/archive/ and it seems it was introduced between Java 18 and 19:

```
$ JAVA_HOME=/tmp/java/jdk-18.0.2 jruby capturePython.rb
JRuby 9.4.5.0, Java 18.0.2
Open3.capture2: stdout io flags: 1 == O_WRONLY
Manual spawn: stdout io flags: 1 == O_WRONLY
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY

$ JAVA_HOME=/tmp/java/jdk-19.0.1 jruby capturePython.rb
JRuby 9.4.5.0, Java 19.0.1
Open3.capture2: stdout io flags: 2049 == O_WRONLY+O_NONBLOCK
Manual spawn: stdout io flags: 2049 == O_WRONLY+O_NONBLOCK
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY
```

I'll see if I can maybe bisect this on the JDK side.

**Comment by headius** [comment]
Yeah, that would be great. I haven't dug into the spawn code yet but I can't imagine anything in there changing with jvm's version. We don't even set non-block as part of that logic. I'm wondering if this has something to do with project loom and virtual threads.

**Comment by iloveeclipse** [comment]
There were few changes related to input/output...
https://www.oracle.com/java/technologies/javase/19all-relnotes.htm

* Charset change for console: https://bugs.openjdk.org/browse/JDK-8283620
* FileChannel.transferFrom May Transfer Fewer Bytes than Expected https://bugs.openjdk.org/browse/JDK-8286763
* FileChannel.lock/tryLock Changed to Treat Size 0 to Mean the Locked Region Goes to End of File https://bugs.openjdk.org/browse/JDK-5041655

Which specific Java API is used by Jruby in this "capturing subprocess" use case?

**Comment by headius** [comment]
@iloveeclipse We literally just use java.nio.channels.Pipe. We do dig the file descriptors out, and possibly set things like O_CLOEXEC but unless requested we should not be setting things nonblock (and certainly not changing that behavior across JVM versions).

Here's where the pipe is created: https://github.com/jruby/jruby/blob/687a89f3b8ff83de2b2c611c904f92f2b657a702/core/src/main/java/org/jruby/util/io/PosixShim.java#L296-L315

**Comment by iloveeclipse** [comment]
The Pipe self isn't changed, but `sun.nio.ch.IOUtil` is changed between 17 and 21, also `java.nio.new JavaNioAccess() {...}.newMappedByteBuffer(UnmapperProxy, long, int, Object, MemorySegment)` implementation from `java.nio.Buffer`. Not sure if that is related, just blindly compared some classes used by Pipe.

One has to bisect that...

**Comment by headius** [comment]
I don't have a setup here for bisecting JDK so if @kschoelhorn can try that I will wait to see results. I do not see anything we are doing in JRuby that would cause this, so far.

**Comment by kschoelhorn** [comment]
@headius I'll give the bisect a try next week! 🙂

**Comment by kschoelhorn** [comment]
@headius The bisect is done and as you already expected is caused by the virtual threads implementation:

```
9583e3657e43cc1c6f2101a64534564db2a9bd84 is the first bad commit
commit 9583e3657e43cc1c6f2101a64534564db2a9bd84
Author: Alan Bateman <alanb@openjdk.org>
Date:   Sat May 7 08:06:16 2022 +0000

    8284161: Implementation of Virtual Threads (Preview)
    
    Co-authored-by: Ron Pressler <rpressler@openjdk.org>
    Co-authored-by: Alan Bateman <alanb@openjdk.org>
    Co-authored-by: Erik Österlund <eosterlund@openjdk.org>
    Co-authored-by: Andrew Haley <aph@openjdk.org>
    Co-authored-by: Rickard Bäckman <rbackman@openjdk.org>
    Co-authored-by: Markus Grönlund <mgronlun@openjdk.org>
    Co-authored-by: Leonid Mesnik <lmesnik@openjdk.org>
    Co-authored-by: Serguei Spitsyn <sspitsyn@openjdk.org>
    Co-authored-by: Chris Plummer <cjplummer@openjdk.org>
    Co-authored-by: Coleen Phillimore <coleenp@openjdk.org>
    Co-authored-by: Robbin Ehn <rehn@openjdk.org>
    Co-authored-by: Stefan Karlsson <stefank@openjdk.org>
    Co-authored-by: Thomas Schatzl <tschatzl@openjdk.org>
    Co-authored-by: Sergey Kuksenko <skuksenko@openjdk.org>
    Reviewed-by: lancea, eosterlund, rehn, sspitsyn, stefank, tschatzl, dfuchs, lmesnik, dcubed, kevinw, amenkov, dlong, mchung, psandoz, bpb, coleenp, smarks, egahlin, mseledtsov, coffeys, darcy
```

Unfortunately, the commit is **huge**: https://github.com/openjdk/jdk/commit/9583e3657e43cc1c6f2101a64534564db2a9bd84 

I looked around a bit in the changes and found two interesting places:

* https://github.com/openjdk/jdk/commit/9583e3657e43cc1c6f2101a64534564db2a9bd84#diff-2d7614743276570c34edc24c52cd078ad30dd31990cb3efda8812cb865cd87c4
* https://github.com/openjdk/jdk/commit/9583e3657e43cc1c6f2101a64534564db2a9bd84#diff-2a2ea179e2d201ad7c2b30fcf1db67e084f7ff975c50a58907d35a56d4c7f19f

It seems that the use of virtual threads puts sockets (and so maybe also pipes?) into non-blocking mode and then high-level code has to emulate blocking behavior on top of that?

**Comment by headius** [comment]
@kschoelhorn Ahh, as I suspected. So my guess here is that this is to more easily support context switching away from a one virtual thread blocked on IO to another virtual thread that's ready to run. We can probably expect this is the case for all "selectable" channels on JVM going forward.

The Ruby world actually made this decision a few releases back for the same reasons: to allow fibers (Ruby equivalent of virtual threads) to be automatically scheduled when making blocking IO calls. In JRuby we did not make that change directly, but we already had been turning the non-block bit on and off every time we make a blocking call, so that we can use `select` and interrupt the operation if necessary. In essence, we left channels blocking but still flip them to non-blocking before reads and writes.

I could not see what you linked; both went to gigantic diffs that did not fully render (most of the diffs collapsed because it's so huge). If you can click on a specific line in that diff I think that would help.

**Comment by kschoelhorn** [comment]
@headius Couldn't make it work with the commit link, so changed them to file links (at the right commit).

**Comment by headius** [comment]
@ioquatix Hey you might be interested in this. Project Loom made a decision similar to CRuby's, to start making some blocking IO nonblocking by default. The links that @kschoelhorn provided above show that they do it upon the first blocking operation with a timeout, or the first blocking operation called from a virtual thread.

**Comment by ioquatix** [comment]
My gut reaction to this is all IO operations should be able to deal with EAGAIN. The fault here lies with the child process. However I can also see how this is breaking behavior if there is insufficient error handling, and that it could be reasonable for the child process to assume blocking behavior. I've certainly run into code like this e.g. `printf` and friends.

**Comment by headius** [comment]
@ioquatix You bring up a good point... @kschoelhorn I would also recommend not trying to set them blocking, since that will cause any virtual threads on JRuby side to start blocking and pin the native thread.

I'm not sure if there's any workaround here or anything to be fixed; in the future all these streams will be set non-blocking to support structured concurrency with fibers/vthreads, so it's inevitable.

**Comment by kschoelhorn** [comment]
@headius The thing is, that using the native `java.lang.ProcessBuilder` in newer Java versions does _still_ create a blocking pipe (or at least the write-side is blocking) for the subprocess (see my example in the OP), not breaking any other programs that currently expect stdout to be blocking.

**Comment by headius** [comment]
> using the native java.lang.ProcessBuilder in newer Java versions does still create a blocking pipe

We do our own native call-out to actually spawn the subprocess, but otherwise just use `java.nio.channels.Pipe` to create the IO channels.

Those pipes are set to non-blocking by a change from the virtual threads drop, in [SinkChannelImpl](https://github.com/openjdk/jdk/blame/master/src/java.base/unix/classes/sun/nio/ch/SinkChannelImpl.java#L82) and [SourceChannelImpl](https://github.com/openjdk/jdk/blob/master/src/java.base/unix/classes/sun/nio/ch/SourceChannelImpl.java#L82).

`ProcessBuilder`, on the other hand, appears to call `pipe()` in native code immediately before forking or spawning, putting the parent-side file descriptors into `ProcessPipe[Input,Output]Stream`, which bottoms out in a `PipeInputStream` (extends `FileInputStream`) or a `FileOutputStream`, neither of which support non-blocking operations.

I'm guessing this has not been changed to maintain backward compatibility. I have argued in the past that Java's subprocess management is awful mostly because there's no way to access a selectable `Channel` for in/out/err, making interactive control basically impossible. The API only returns streams, and since there's no selectable/nonblocking `[Input,Output]Stream` classes in JDK, there's no way to make those streams non-blocking for consumers.

My investigation confirms what you observed: normal Java process launching still uses normal blocking pipes, stuffed into file-oriented streams and channels. This maintains the behavior you're hoping to see, but also means that blocking IO operations on `ProcessBuilder` spawns will be incompatible with Loom and virtual threads. It's a choice, but not one I necessarily agree with.

I came up with a workaround below, which will force the pipes back to blocking before the spawn. Like the `ProcessBuilder` case, this will break the ability to automatically context-switch a blocking vthread (i.e. calls from a `Fiber`), but perhaps you don't care about that case.

There are two oddities in this workaround:

* We must first set nonblock because the Java and JRuby objects wrapping the stream do not know it is already non-blocking. (#8082)
* Due to a bug in JRuby's `IO#fcntl` that does the **opposite** of what you want when setting/clearing `O_NONBLOCK`, I had to reverse the way that flag is set and cleared. (#8081

```ruby
...
require 'fcntl'
flags = write.fcntl(Fcntl::F_GETFL)
write.fcntl(Fcntl::F_SETFL, flags & (~Fcntl::O_NONBLOCK))
write.fcntl(Fcntl::F_SETFL, flags | (Fcntl::O_NONBLOCK))
```

This fixes the output for spawned processes in your original example:

```
$ jruby capturePython.rb 
JRuby 9.4.6.0-SNAPSHOT, Java 21
Open3.capture2: stdout io flags: 5 == O_WRONLY+O_NONBLOCK
Manual spawn: stdout io flags: 1 == O_WRONLY
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY
```

It may not be possible to extend this workaround into the specialized `Open3` functions, since they handle their own pipes. And the same restriction applies here as with `ProcessBuilder`: blocking IO against these pipes will pin vthreads/fibers.

**Comment by ioquatix** [comment]
Very few programs will use the ends of the pipe provided to the child process, normal course of action is to immediately close the remote end (child process end). So, "blocking IO against these pipes will pin vthreads/fibers", while strictly true, probably doesn't matter in practice, as you are modifying the end of the pipe that's given to the child process. The end you will read or write to, you still want to be non-blocking. But that's not provided to the child process.

**Comment by headius** [comment]
I think I've convinced the loom folks that this is a bug.

They confirmed this was done intentionally for the same reasons as CRuby. However they do it lazily; when the first vthread touches a socket, for example, the socket is set to nonblock from then on and blocking calls are simulated with a select loop.

The bug is that they do this **eagerly** for pipes. This is a known issue and there's a patch to make pipes also lazily nonblock, but it's still being debated. This would fix our issue; the child side of the pipe would not be used during process launch, so it would remain nonblocking.

I will be making our case when I talk to the Loom folks at FOSDEM this weekend.

**Comment by headius** [comment]
@kschoelhorn So here's where we stand on this:

https://mail.openjdk.org/pipermail/loom-dev/2024-January/006412.html

The Loom folks seem like they will make pipe streams lazily set O_NONBLOCK like other selectable streams, so I expect this behavior will be fixed in JDK update releases over the next few months.

The best workaround for now will probably be for you to tweak the blocking status of the stream yourself, using the workaround example I showed above. Unfortunately that will not fix this in open3, since that's a separate gem that handles its own pipes.

Another option would be for us to tweak JRuby to set the child side of pipes to nonblocking when they are being passed to spawn... but I'm reluctant to do that because nonblock might be *intentional* and we would be resetting it in error.

**Comment by headius** [comment]
FYI #8081 is fixed for next 9.3 and 9.4 releases, so when those are out you won't have to invert the flag logic for O_NONBLOCK in fcntl.

**Comment by headius** [comment]
@kschoelhorn Thanks for that link! I'm glad they agreed to do it!

**Comment by headius** [comment]
There's not really a good fix we can do in JRuby, since it's due to the JDK eagerly setting pipe streams to non-blocking. We can't simply reset them to blocking because we can't detect whether the JDK did it or the user did it.

Closing as Won't Fix. Workarounds are discussed above: use JDK 17 until this is fixed in releases of 18+, or manually reset the pipes to blocking yourself.

**Comment by iloveeclipse** [comment]
Links to related tickets:
- Fix in Java 23+:  https://bugs.openjdk.org/browse/JDK-8325028
- Backport request to 21 (on RH tracker): https://issues.redhat.com/browse/OPENJDK-2813

If someone can create backport request to Java 21 at bugs.openjdk.org, please don't hesitate to do so :-)

**Comment by headius** [comment]
PR that fixed this is here: https://github.com/openjdk/jdk/pull/17648

I will look into making a backport request on JBS.

**Comment by headius** [comment]
I created https://bugs.openjdk.org/browse/JDK-8329604 to request backport. Hopefully I did it right 🤷

**Comment by iloveeclipse** [comment]
@headius : **many** thanks!

**Comment by kschoelhorn** [comment]
This is now fixed on Java 21 with the new 21.0.4! :partying_face: 
```
$ JAVA_HOME=/tmp/jdk/jdk-21.0.4+7 jruby capturePython.rb
JRuby 9.2.21.0, Java 21.0.4
Open3.capture2: stdout io flags: 1 == O_WRONLY
Manual spawn: stdout io flags: 1 == O_WRONLY
Backticks: stdout io flags: 1 == O_WRONLY
java.lang.ProcessBuilder: stdout io flags: 1 == O_WRONLY
```
Thanks everyone again for getting this fixed so quickly!

**Comment by headius** [comment]
@kschoelhorn Thanks for the update!
